File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/Java/C/s514557145_comments_removed.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define MOD2 998244353
#define MAX_N 100000

long long tenmod[MAX_N + 1];
long long inv9;

typedef struct {
    long long sum;
    int l;
    int r;
} S;

const S S_E = {0, INT_MAX / 2, INT_MIN / 2};

typedef struct {
    long long D;
    int N;
} F;

const F F_I = {0, 0};

long long modInv(long long x) {
    return modPow(x, MOD2 - 2);
}

long long modPow(long long x, long long y) {
    long long res = 1;
    while (y > 0) {
        if (y % 2 == 1) {
            res = (res * x) % MOD2;
            if (res < 0) res += MOD2;
        }
        x = (x * x) % MOD2;
        if (x < 0) x += MOD2;
        y /= 2;
    }
    return res;
}

S S_op(S s1, S s2) {
    long long sum = (s1.sum + s2.sum) % MOD2;
    if (sum < 0) sum += MOD2;
    int new_l = (s1.l < s2.l) ? s1.l : s2.l;
    int new_r = (s1.r > s2.r) ? s1.r : s2.r;
    return (S) {sum, new_l, new_r};
}

S S_map(F f, S s) {
    int n_minus_l = f.N - s.l;
    int n_minus_r = f.N - s.r - 1;
    long long c = (tenmod[n_minus_l] - tenmod[n_minus_r]);
    if (c < 0) c += MOD2;
    long long new_sum = (f.D * c) % MOD2;
    new_sum = (new_sum * inv9) % MOD2;
    return (S) {new_sum, s.l, s.r};
}

F F_composite(F f, F g) {
    return (F) {f.D, f.N};
}

typedef struct {
    int MAX;
    int N;
    int Log;
    S (*Op)(S, S);
    S E;
    S (*Mapping)(F, S);
    F (*Composition)(F, F);
    F Id;
    S *Dat;
    F *Laz;
} LazySegTree;

void push(LazySegTree *tree, int k) {
    F f = tree->Laz[k];
    if (f.D == F_I.D && f.N == F_I.N) return;
    int lk = (k << 1);
    int rk = (k << 1) + 1;
    tree->Dat[lk] = (*tree->Mapping)(f, tree->Dat[lk]);
    tree->Dat[rk] = (*tree->Mapping)(f, tree->Dat[rk]);
    if (lk < tree->N) {
        tree->Laz[lk] = (*tree->Composition)(f, tree->Laz[lk]);
    }
    if (rk < tree->N) {
        tree->Laz[rk] = (*tree->Composition)(f, tree->Laz[rk]);
    }
    tree->Laz[k] = F_I;
}

void pushTo(LazySegTree *tree, int lk, int rk) {
    int log_n = tree->Log;
    for (int i = log_n; i > 0; i--) {
        int k_l = lk >> i;
        if (((k_l << i) != lk)) {
            push(tree, k_l);
        }
        int k_r = rk >> i;
        if (((k_r << i) != rk)) {
            push(tree, k_r);
        }
    }
}

void updateFrom(LazySegTree *tree, int lk, int rk) {
    int log_n = tree->Log;
    for (int i = 1; i <= log_n; i++) {
        int lki = lk >> i;
        if (((lki << i) != lk)) {
            tree->Dat[lki] = (*tree->Op)(tree->Dat[lki * 2], tree->Dat[lki * 2 + 1]);
        }
        int rki = (rk - 1) >> i;
        if (((rki << i) != (rk - 1))) {
            tree->Dat[rki] = (*tree->Op)(tree->Dat[rki * 2], tree->Dat[rki * 2 + 1]);
        }
    }
}

LazySegTree *LazySegTree_init(S *dat, int dat_len, S (*op)(S, S), S E, S (*mapping)(F, S), F (*composition)(F, F), F Id) {
    LazySegTree *tree = (LazySegTree*)malloc(sizeof(LazySegTree));
    int n = 1;
    while (n < dat_len) n <<= 1;
    tree->MAX = dat_len;
    tree->N = n;
    tree->Log = (int)log2(n);
    tree->Op = op;
    tree->E = E;
    tree->Mapping = mapping;
    tree->Composition = composition;
    tree->Id = Id;
    tree->Dat = (S*)malloc((tree->N << 1) * sizeof(S));
    tree->Laz = (F*)malloc(tree->N * sizeof(F));
    for (int i = 0; i < (tree->N << 1); i++) {
        tree->Dat[i] = E;
    }
    for (int i = 0; i < tree->N; i++) {
        tree->Laz[i] = Id;
    }
    for (int i = 0; i < dat_len; i++) {
        tree->Dat[tree->N + i] = dat[i];
    }
    for (int i = tree->N - 1; i > 0; i--) {
        tree->Dat[i] = (*op)(tree->Dat[i * 2], tree->Dat[i * 2 + 1]);
    }
    return tree;
}

void LazySegTree_apply(LazySegTree *tree, int l, int r, F f) {
    if (l >= r) return;
    int N = tree->N;
    l += N;
    r += N;
    pushTo(tree, l, r);
    for (int l2 = l, r2 = r; l2 < r2; ) {
        if ((l2 & 1) == 1) {
            tree->Dat[l2] = (*tree->Mapping)(f, tree->Dat[l2]);
            if (l2 < N) {
                tree->Laz[l2] = (*tree->Composition)(f, tree->Laz[l2]);
            }
            l2++;
        }
        if ((r2 & 1) == 1) {
            r2--;
            tree->Dat[r2] = (*tree->Mapping)(f, tree->Dat[r2]);
            if (r2 < N) {
                tree->Laz[r2] = (*tree->Composition)(f, tree->Laz[r2]);
            }
        }
        l2 >>= 1;
        r2 >>= 1;
    }
    updateFrom(tree, l, r);
}

int main() {
    int N, Q;
    scanf("%d %d", &N, &Q);
    inv9 = modInv(9);
    tenmod[0] = 1;
    for (int j = 1; j <= N; j++) {
        tenmod[j] = (tenmod[j - 1] * 10LL) % MOD2;
    }
    S *dat = (S*)malloc(N * sizeof(S));
    for (int i = 0; i < N; i++) {
        dat[i].sum = tenmod[N - i - 1];
        dat[i].l = i;
        dat[i].r = i;
    }
    LazySegTree *tree = LazySegTree_init(dat, N, S_op, S_E, S_map, F_composite, F_I);
    for (int j = 0; j < Q; j++) {
        int l, r;
        long long D;
        scanf("%d %d %lld", &l, &r, &D);
        l--;
        r--;
        F f = {D, N};
        LazySegTree_apply(tree, l, r + 1, f);
        printf("%lld\n", tree->Dat[1].sum);
    }
    free(dat);
    free(tree->Dat);
    free(tree->Laz);
    free(tree);
    return 0;
}
// End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/Java/Python/s514557145_comments_removed.py
class S:
    E = None

    def __init__(self, sum_val, l, r):
        self.sum = sum_val % MOD2
        self.l = l
        self.r = r

    @staticmethod
    def op(s1, s2):
        new_sum = (s1.sum + s2.sum) % MOD2
        new_l = min(s1.l, s2.l)
        new_r = max(s1.r, s2.r)
        return S(new_sum, new_l, new_r)

    @staticmethod
    def map(f, s):
        N = f.N
        term1 = tenmod[N - s.l]
        term2 = tenmod[N - s.r - 1]
        c = (term1 - term2) % MOD2
        new_sum = (f.D * c) % MOD2
        new_sum = (new_sum * inv9) % MOD2
        return S(new_sum, s.l, s.r)

S.E = S(0, float('inf'), float('-inf'))

class F:
    I = None

    def __init__(self, D, N):
        self.D = D
        self.N = N

    @staticmethod
    def composite(f, g):
        return F(f.D, f.N)

F.I = F(0, 0)

class LazySegTree:
    def __init__(self, dat, op_func, e, mapping_func, composition_func, id_f):
        self.N = 1
        self.MAX = len(dat)
        while self.N < self.MAX:
            self.N <<= 1
        self.Log = (self.N).bit_length() - 1
        self.Op = op_func
        self.E = e
        self.Mapping = mapping_func
        self.Composition = composition_func
        self.Id_f = id_f

        self.Dat = [self.E] * (2 * self.N)
        self.Laz = [self.Id_f] * self.N

        for i in range(self.MAX):
            self.Dat[self.N + i] = dat[i]
        for i in range(self.N - 1, 0, -1):
            self.Dat[i] = self.Op(self.Dat[2 * i], self.Dat[2 * i + 1])

    def allProd(self):
        return self.Dat[1].sum

    def apply(self, l, r, f):
        if l >= r:
            return
        l_internal = l + self.N
        r_internal = r + self.N
        self.pushTo(l_internal, r_internal)
        l2 = l_internal
        r2 = r_internal
        while l2 < r2:
            if (l2 & 1) == 1:
                self.Dat[l2] = self.Mapping(f, self.Dat[l2])
                if l2 < self.N:
                    self.Laz[l2] = self.Composition(f, self.Laz[l2])
                l2 += 1
            if (r2 & 1) == 1:
                r2 -= 1
                self.Dat[r2] = self.Mapping(f, self.Dat[r2])
                if r2 < self.N:
                    self.Laz[r2] = self.Composition(f, self.Laz[r2])
            l2 >>= 1
            r2 >>= 1
        self.updateFrom(l_internal, r_internal)

    def pushTo(self, l, r):
        for i in range(self.Log, 0, -1):
            l_node = l >> i
            if (l_node << i) != l:
                self.push(l_node)
            r_node = (r - 1) >> i
            if (r_node << i) != (r - 1):
                self.push(r_node)

    def push(self, k):
        if self.Laz[k] == self.Id_f:
            return
        left = 2 * k
        right = 2 * k + 1
        self.Dat[left] = self.Mapping(self.Laz[k], self.Dat[left])
        self.Dat[right] = self.Mapping(self.Laz[k], self.Dat[right])
        if left < self.N:
            self.Laz[left] = self.Composition(self.Laz[k], self.Laz[left])
            self.Laz[right] = self.Composition(self.Laz[k], self.Laz[right])
        self.Laz[k] = self.Id_f

    def updateFrom(self, l, r):
        for i in range(1, self.Log + 1):
            l_node = l >> i
            if (l_node << i) != l:
                self.Dat[l_node] = self.Op(self.Dat[2 * l_node], self.Dat[2 * l_node + 1])
            r_node = (r - 1) >> i
            if (r_node << i) != (r - 1):
                self.Dat[r_node] = self.Op(self.Dat[2 * r_node], self.Dat[2 * r_node + 1])

MOD2 = 998244353
inv9 = 0

def modPow(x, y):
    res = 1
    while y > 0:
        if y % 2 == 1:
            res = (res * x) % MOD2
        x = (x * x) % MOD2
        y = y // 2
    return res

def modInv(x):
    return modPow(x, MOD2 - 2)

inv9 = modInv(9)

def main():
    import sys
    input = sys.stdin.read().split()
    ptr = 0
    N = int(input[ptr]); ptr +=1
    Q = int(input[ptr]); ptr +=1

    global tenmod
    tenmod = [0] * (N + 1)
    tenmod[0] = 1
    for j in range(1, N + 1):
        tenmod[j] = (tenmod[j-1] * 10) % MOD2

    dat = []
    for i in range(N):
        coeff = tenmod[N - i - 1]
        s = S(coeff, i, i)
        dat.append(s)

    lazySegTree = LazySegTree(dat, S.op, S.E, S.map, F.composite, F.I)

    for _ in range(Q):
        l = int(input[ptr])-1; ptr +=1
        r = int(input[ptr])-1; ptr +=1
        D = int(input[ptr]); ptr +=1
        f = F(D, N)
        lazySegTree.apply(l, r+1, f)
        print(lazySegTree.allProd())

if __name__ == "__main__":
    main()
#End of Code
